sdk/eventhub/Azure.Messaging.EventHubs/tests/Processor/EventProcessorLiveTests.cs

```diff
@@ -0,0 +1,907 @@
+                                }
+                        );
+
+                    hub.AddEventProcessors(1);
+
+                    // Start the event processors.
+
+                    await hub.StartAllAsync();
+
+                    // Make sure the event processors have enough time to stabilize.
+                    // TODO: we'll probably need to extend this delay once load balancing is implemented.
+
+                    await Task.Delay(5000);
+
+                    // CloseAsync should have not been called when constructing the event processor or initializing the partition processors.
+
+                    Assert.That(closeCalls.Keys, Is.Empty);
+
+                    // Stop the event processors.
+
+                    await hub.StopAllAsync();
+
+                    // Validate results.
+
+                    var partitionIds = await client.GetPartitionIdsAsync();
+
+                    foreach (var partitionId in partitionIds)
+                    {
+                        Assert.That(closeCalls.TryGetValue(partitionId, out var calls), Is.True, $"{ partitionId }: CloseAsync should have been called.");
+                        Assert.That(calls, Is.EqualTo(1), $"{ partitionId }: CloseAsync should have been called only once.");
+
+                        Assert.That(closeReasons.TryGetValue(partitionId, out var reason), Is.True, $"{ partitionId }: close reason should have been set.");
+                        Assert.That(reason, Is.EqualTo(PartitionProcessorCloseReason.Shutdown), $"{ partitionId }: unexpected close reason.");
+                    }
+
+                    Assert.That(closeCalls.Keys.Count, Is.EqualTo(partitionIds.Count()));
+                }
+            }
+        }
+
+        /// <summary>
+        ///   Verifies that the <see cref="EventProcessor" /> is able to
+        ///   connect to the Event Hubs service and perform operations.
+        /// </summary>
+        ///
+        [Test]
+        public async Task PartitionProcessorProcessEventsAsyncReceivesAllEvents()
+        {
+            await using (var scope = await EventHubScope.CreateAsync(2))
+            {
+                var connectionString = TestEnvironment.BuildConnectionStringForEventHub(scope.EventHubName);
+
+                await using (var client = new EventHubClient(connectionString))
+                {
+                    var allReceivedEvents = new ConcurrentDictionary<string, List<EventData>>();
+
+                    // Create the event processor hub to manage our event processors.
+
+                    var hub = new EventProcessorHub
+                        (
+                            EventHubConsumer.DefaultConsumerGroupName,
+                            client,
+                            onProcessEvents: (partitionContext, checkpointManager, events, cancellationToken) =>
+                            {
+                                if (events.Count() > 0)
```

Comment:
Given your pattern, it may make sense to refactor to something like:

```csharp

var partitionEvents = new List<EventData>(events ?? Enumerable.Empty<EventData>());

if (partitionEvents.Count > 0)
{
    allReceivedEvents.AddOrUpdate
    (
        partitionContext.PartitionId,
        partitionId => partitionEvents,
        (partitionId, list) =>
        {
            list.AddRange(events);
            return list;
        }
    );
```